src/fork_gc.c

```diff
@@ -95,26 +97,25 @@ static void FGC_sendTerminator(ForkGC *fgc) {
 
 static long long FGC_recvLongLong(ForkGC *fgc) {
   long long ret;
-  ssize_t sizeRead = read(fgc->pipefd[GC_READERFD], &ret, sizeof(ret));
-  if (sizeRead != sizeof(ret)) {
-    return 0;
-  }
+  FGC_recvFixed(fgc, &ret, sizeof ret);
   return ret;
 }
 
 static void *FGC_recvPtrAddr(ForkGC *fgc) {
   void *ret;
-  ssize_t sizeRead = read(fgc->pipefd[GC_READERFD], &ret, sizeof(ret));
-  if (sizeRead != sizeof(ret)) {
-    return 0;
-  }
+  FGC_recvFixed(fgc, &ret, sizeof ret);
   return ret;
 }
 
 static void FGC_recvFixed(ForkGC *fgc, void *buf, size_t len) {
-  ssize_t nrecvd = read(fgc->pipefd[GC_READERFD], buf, len);
-  if (nrecvd != len) {
-    printf("warning: got a bad length when writing to pipe.\r\n");
+  while (len) {
+    ssize_t nrecvd = read(fgc->pipefd[GC_READERFD], buf, len);
+    if (nrecvd > 0) {
+      buf += nrecvd;
+      len -= nrecvd;
+    } else if (nrecvd < 0) {
+      printf("Got error while reading from pipe (%s)", strerror(errno));
```

Comment:
We need to somehow have a clean exit here, maybe we can use longjump?
Also I suggest to write it using RedisModule_Log() (Since redis 5.0.6 I think we will see logs from threadSafeCtx).

Reply:
We can just break from the loop. I guess we should also check each read to ensure that we're able to read from it successfully. setjmp/longjmp etc. will result in leaks.

Reply:
You do not just need to break, you also need to expose the failure to the caller to handle it.